src/kernels/jupyter/finder/remoteKernelFinder.ts

```diff
@@ -67,102 +91,128 @@ export class RemoteKernelFinder implements IRemoteKernelFinder, IExtensionSingle
-                    traceError('Failed to get remote kernels', ex);
-                }
-                return [];
+        this.disposables.push(
+            this.serverUriStorage.onDidChangeUri(() => {
+                this.updateCache().then(noop, noop);
+            })
+        );
+
+        // If we create a new kernel, we need to refresh if the kernel is remote (because
+        // we have live sessions possible)
+        // Note, this is a perf optimization for right now. We should not need
+        // to check for remote if the future when we support live sessions on local
+        this.kernelProvider.onDidStartKernel((k) => {
+            if (isRemoteConnection(k.kernelConnectionMetadata)) {
+                // update remote kernels
+                this.updateCache().then(noop, noop);
             }
+        });
+
+        // For kernel dispose we need to wait a bit, otherwise the list comes back the
+        // same
+        this.kernelProvider.onDidDisposeKernel(
+            (k) => {
+                if (k && isRemoteConnection(k.kernelConnectionMetadata)) {
+                    setTimeout(() => {
+                        this.updateCache().then(noop, noop);
+                    }, REMOTE_KERNEL_REFRESH_INTERVAL);
+                }
+            },
+            this,
+            this.disposables
         );
 
-        traceVerbose(`KernelFinder discovered ${kernels.length} remote`);
-        return kernels;
+        this.extensions.onDidChange(
+            () => {
+                // If we just installed the Python extension and we fetched the controllers, then fetch it again.
+                if (
+                    !this.wasPythonInstalledWhenFetchingControllers &&
+                    this.extensionChecker.isPythonExtensionInstalled
+                ) {
+                    this.updateCache().then(noop, noop);
+                }
+            },
+            this,
+            this.disposables
+        );
+        this.wasPythonInstalledWhenFetchingControllers = this.extensionChecker.isPythonExtensionInstalled;
     }
 
-    private async listKernelsImpl(
-        resource: Resource,
-        cancelToken: CancellationToken | undefined,
-        useCache: 'ignoreCache' | 'useCache'
-    ) {
+    private async loadCache() {
         if (this.serverConnectionType.isLocalLaunch) {
-            return [];
+            await this.writeToCache([]);
+            this._initializeResolve();
+            return;
         }
 
-        const kernelsFromCachePromise =
-            useCache === 'ignoreCache' ? Promise.resolve([]) : this.getFromCache(cancelToken);
-        let updateCache = true;
-        const kernelsWithoutCachePromise = (async () => {
-            const connInfo = await this.getRemoteConnectionInfo(cancelToken);
-            return connInfo ? this.listKernelsFromConnection(resource, connInfo) : Promise.resolve([]);
-        })();
+        const loadCancellationToken = new CancellationTokenSource();
```

Comment:
I'll confess that I might just be ignorant on the cancellation tokens (I know I've gotten mixed on them before). But don't they have to be passed down or saved and hooked up to something to actually perform the cancellation? From my quick reading this didn't look to me like it was hooked up to anything that could actually cancel the load operation.

Reply:
This should be cleaned up. For single operation, we want to drop the cancellation token and for operations can be performed multiple times, we want to reuse the cancellation token to dispose previous requests. Will update soon.